File: pa4/List.c
//PA2
//CMPS 101 Tantalo Spring 2019
//Aaron Nguyen
//anguy200
//1585632








#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include "List.h"

typedef struct NodeObj{
  int data;
  struct NodeObj *next;
  struct NodeObj *prev;
} NodeObj;

typedef NodeObj *Node;

Node newNode(int data){
  Node N = malloc(sizeof(NodeObj));
  assert(N != NULL);
  N->data = data;
  N->next = NULL;
  return(N);
}

void freeNode(Node *pN){
  if(pN != NULL && *pN != NULL){
    free(*pN);
    *pN = NULL;
  }
}

typedef struct ListObj{
  Node head;
  Node tail;
  Node cursor;
  int index;
  int numItems;
} ListObj;

typedef struct ListObj *List;

List newList(void){
  List L = malloc(sizeof(ListObj));
  assert(L!=NULL);
  L->head = NULL;
  L->tail = NULL;
  L->cursor = NULL;
  L-> index = -1;
  L-> numItems = 0;
  return L;
}

void freeList(List *pL){
  if(pL != NULL && *pL != NULL){
    clear(*pL);
    free(*pL);
    *pL = NULL;
  }
}


int length(List L){
  if(L == NULL){
    printf("This shit broke");
  }
  return L->numItems;
}

int index(List L){
  return L->index;
}

int front(List L){
  if(L->numItems > 0){
    return L->head->data;
  }else{
    return -1;
  }
}

int back(List L){
  if(L->numItems > 0){
    return L->tail->data;
  }else{
    return -1;
  }
}

int get(List L){
  if(L->numItems > 0 && L->index >= 0 && L->index < L->numItems){
    return L->cursor->data;
  }else{
    return -1;
  }
}

int equals(List A, List B){
  if(A->numItems != B->numItems){
    return 0;
  }else{
    Node tempA = A->head;
    Node tempB = B->head;
    for(int x = 0; x < A->numItems; x++){
      if(tempA->data != tempB->data){
        return 0;
      }else{
        tempA = tempA->next;
        tempB = tempB->next;
      }
    }
    return 1;
  }
}

void clear(List L){
  Node N = NULL;
  if( L == NULL ){
   perror("List is Null or empty");
   exit(EXIT_FAILURE);
  }
  L->cursor = NULL;
  L->index = -1;
  while(L->numItems > 0){
    N = L-> head;
    L->head = L->head->next;
    N->next = NULL;
    freeNode(&N);
    L->numItems--;
  }
}

void moveFront(List L){
  if(L->numItems == 0){
    perror("List is Empty");
    exit(EXIT_FAILURE);
  }else{
    L->cursor = L->head;
    L->index = 0;
  }
}

void moveBack(List L){
  if(L-> numItems == 0){
    perror("List is Empty");
    exit(EXIT_FAILURE);
  }else{
    L->cursor = L->tail;
    L->index = L->numItems - 1;
  }
}

void movePrev(List L){
  if(L->index != -1 && L->cursor != L->head){
    L->cursor = L->cursor->prev;
    L->index--;
  }else if(L->cursor == L->head){
    L->cursor = NULL;
    L->index = -1;
  }
}

void moveNext(List L){
  if(L->cursor != NULL && L->cursor != L->tail){
    L->cursor = L->cursor->next;
    L->index++;
  }else if(L->cursor == L->tail){
    L->cursor = NULL;
    L->index = -1;
  }
}


void prepend(List L, int data){
  //Node N = NULL;
  //Node END = NULL;
  if(L==NULL){
    perror("List was not created properly");
    exit(EXIT_FAILURE);
  }
  if(L->numItems == 0){
    Node N = newNode(data);
    L->head = L->tail = N;
  }else{
    Node temp = L->head;
    L-> head = newNode(data);
    L->head->next = temp;
    temp->prev = L->head;
    if(L->index != -1){
      L->index++;
    }
    
  }
  L->numItems++;

}

void append(List L, int data){
  if(L==NULL){
    perror("List was not created correctly");
    exit(EXIT_FAILURE);
  }if(L->numItems == 0){
    L->head =  newNode(data);
    L->tail = L->head;
  }else{
    L->tail->next = newNode(data);
    L->tail->next->prev = L->tail;
    L->tail = L->tail->next;
  }
  L->numItems++;
}

void insertBefore(List L, int data){
  if(L->numItems != 0 && L->index >= 0){
    if(L->index == 0){
      prepend(L, data);
    }else{
      Node A = L->cursor->prev;
      Node temp = newNode(data);
      A->next = temp;
      temp -> prev = A;
      temp -> next = L->cursor;
      L->cursor->prev = temp;
      L->index++;
      L->numItems++;
    }
  }
}

void insertAfter(List L, int data){
  if(L->numItems != 0 && L->numItems >= 0){
    if(L->cursor == L->tail){
      append( L, data);
    }else{
      Node A = L->cursor->next;
      Node temp = newNode(data);
      A->prev = temp;
      temp->next = A;
      temp->prev = L->cursor;
      L->cursor->next = temp;
      L->numItems++;
    }
  }
}

void deleteFront(List L){
  if(L->numItems != 0){
    Node temp = L->head;
    L->head = L-> head->next;
    L->head->prev = NULL;
    freeNode(&temp);
    L->numItems--;
    if(L->index == 0){
      L->index--;
      L->cursor = NULL;
    }else if(L-> index != -1){
      L->index--;
    }
  }
}

void deleteBack(List L){
  if(L->numItems > 0){
    if(L->cursor == L->tail){
      L->cursor = NULL;
      L->index = -1;
    }if(L->numItems == 1){
      Node temp = L->tail;
      freeNode(&temp);
      L->tail = NULL;
      L->head = NULL;
      L->numItems--;
    }else{
      Node temp = L->tail;
      L->tail = L->tail->prev;
      L->tail->next = NULL;
      freeNode(&temp);
      L->numItems--;
    }
  }

}

void delete(List L){
  if(L->index == 0 && L->numItems > 0){
    deleteFront(L);
  }else if(L-> index == L->numItems - 1 && L -> numItems > 0){
    deleteBack(L);
  }else if(L->index >= 0 && L->numItems !=0){
    Node curprev = L->cursor->prev;
    Node curnext = L->cursor->next;
    curprev->next = curnext;
    curnext->prev = curprev;
    L->numItems--;
    freeNode(&L->cursor);
    L->index = -1;
    L->cursor = NULL;
  }

}

int dequeue(List L){
  if(L != NULL){
    moveFront(L);
    int x = get(L);
    delete(L);
    return x;
  }
  return -1;
}

void printNode(FILE *out, Node N){
  fprintf(out, "%d ", N->data);
}

void printList(FILE* out, List L){
  Node N = NULL;
  if(L==NULL){
    perror("List is not initialized");
    exit(EXIT_FAILURE);
  }
  N = L->head;
  while(N!=NULL){
    fprintf(out, "%d ", N->data);
    N = N->next;
  }
}

List copyList(List L){
  List tempList = newList();
  Node temp = L->head;
  while(temp != NULL){
    append(tempList, temp->data);
    temp = temp->next;
  }
  return tempList;
}

File: pa1/List.java
import java.lang.String;
import java.lang.IndexOutOfBoundsException;

public class List {
    private class Node{
        int data;
        Node next;
        Node prev;

        Node(int x){
            data = x;
            next = null;
            prev = null;
        }
    }

    private Node head;
    private Node tail;
    private Node cursor;
    private int cursorInd;
    private int numItems;

    public List(){
        head = tail = cursor =null;
        cursorInd = -1;
        numItems = 0;
    }

    public int length(){
        return numItems;
    }

    public int index(){
        return cursorInd;
    }

    public int front() throws IndexOutOfBoundsException{
        if(numItems > 0) {
            return head.data;
        }else{
            throw new IndexOutOfBoundsException("list is empty");
        }
    }

    public int back(){
        if(numItems > 0) {
            return tail.data;
        }else{
            throw new IndexOutOfBoundsException("list is empty");
        }
    }

    public int get(){
        if(numItems > 0 && cursorInd >= 0 && cursorInd < numItems ) {
            return cursor.data;
        }else{
            return -1;
        }
    }

    public boolean equals(List L){
        if(numItems != L.numItems){
            return false;
        }else{
            Node temp = head;
            Node tempL = L.head;
            for(int x = 0; x < numItems; x++){
                if(temp.data != tempL.data){
                    return false;
                }else{
                    temp = temp.next;
                    tempL = tempL.next;
                }
            }
            return true;
        }
    }
    public void clear(){
        head = tail = cursor = null;
        numItems = 0;
        cursorInd = -1;
    }
    public void moveFront() throws IndexOutOfBoundsException{
        if(numItems == 0){
            //throw new IndexOutOfBoundsException("List is Empty");
        }else{
            cursor = head;
            cursorInd = 0;
        }
    }

    public void moveBack(){
        if(numItems == 0){
            //throw new IndexOutOfBoundsException("List is Empty");
        }else{
            cursor = tail;
            cursorInd = numItems - 1;
        }
    }

    public void movePrev(){
        if(cursorInd != -1 && cursor != head){
            cursor = cursor.prev;
            cursorInd--;
        }else if(cursor == head){
            cursor = null;
            cursorInd = -1;
        }
    }

    public void moveNext(){
        if(cursor != null && cursor != tail){
            cursor = cursor.next;
            cursorInd++;
        }else if(cursor == tail){
            cursor = null;
            cursorInd = -1;
        }
    }

    public void prepend(int data){
        if(numItems == 0){
            head = tail = new Node(data);
            numItems++;
        }else{
            Node temp = head;
            head = new Node(data);
            head.next = temp;
            temp.prev = head;
            numItems++;
            if(cursorInd != -1 ){
                cursorInd++;
            }
        }
    }

    public void append(int data){
        if(numItems == 0){
            head = new Node(data);
            tail = head;
            numItems++;
        }else{
            tail.next = new Node(data);
            tail.next.prev = tail;
            tail = tail.next;
            numItems++;
        }
    }

    public void insertBefore(int data){
        if(numItems !=0 && cursorInd >= 0){
            if(cursorInd == 0){
                prepend(data);
            }else{
                Node A = cursor.prev;
                Node temp = new Node(data);
                A.next = temp;
                temp.prev = A;
                temp.next = cursor;
                cursor.prev = temp;
                cursorInd++;
                numItems++;
            }
        }
    }

    public void insertAfter(int data){
        if(numItems !=0 && cursorInd >= 0){
            if(cursor == tail){
                append(data);
            }else{
                Node A = cursor.next;
                Node temp = new Node(data);
                A.prev = temp;
                temp.next = A;
                temp.prev = cursor;
                cursor.next = temp;
                numItems++;
            }
        }
    }

    public void deleteFront(){
        if(numItems != 0){
            head = head.next;
            head.prev = null;
            numItems--;
            if(cursorInd == 0){
                cursorInd--;
                cursor = null;
            }else if(cursorInd != -1 ){
                cursorInd--;
            }
        }
    }

    public void deleteBack(){
        if(numItems > 0){
            if (cursor == tail){
                cursor = null;
                cursorInd = -1;
            }
            if(numItems == 1){
                tail = null;
                head = null;
                numItems--;
            }else {
                tail = tail.prev;
                tail.next = null;
                numItems--;
            }
        }
    }
    void delete(){
        if(cursorInd == 0 && numItems > 0){
            deleteFront();
        }else if(cursorInd == numItems - 1 && numItems > 0){
            deleteBack();
        }else if(cursorInd >= 0 && numItems != 0){
            Node curprev  = cursor.prev;
            Node curnext = cursor.next;
            curprev.next = curnext;
            curnext.prev = curprev;
            numItems--;
            cursorInd = -1;
            cursor = null;
        }
    }
    private String myString(Node H){
        if( H==null ){
            return "";
        }else{
            return (H.data+ " " +myString(H.next));
        }
    }

    public String toString(){
        return myString(head);
    }

    public List copy(){
        List A = new List();
        Node temp = head;
        while(temp != null){
            A.append(temp.data);
            temp = temp.next;
        }
        return A;
    }

    public List concat(List L){
        List A = new List();
        Node temp = L.head;
        A = copy();
        while(temp != null){
            A.append(temp.data);
            temp = temp.next;
        }
        return A;
    }
    /*public static void main(String[] args){
        List A = new List();
        List B = new List();

        for(int i=1; i<=20; i++){
            A.append(i);
            B.prepend(i);
        }
        System.out.println(A);
        System.out.println(B);

        for(A.moveFront(); A.index()>=0; A.moveNext()){
            System.out.print(A.get()+" ");
        }
        System.out.println();
        for(B.moveBack(); B.index()>=0; B.movePrev()){
            System.out.print(B.get()+" ");
        }
        System.out.println();

        List C = A.copy();
        System.out.println(A.equals(B));
        System.out.println(B.equals(C));
        System.out.println(C.equals(A));
        System.out.println(C);
        System.out.println(A);

        A.moveFront();
        for(int i=0; i<5; i++) A.moveNext(); // at index 5
        A.insertBefore(-1);                  // at index 6
        for(int i=0; i<9; i++) A.moveNext(); // at index 15
        A.insertAfter(-2);
        for(int i=0; i<5; i++) A.movePrev(); // at index 10
        A.delete();
        System.out.println(A);
        System.out.println(A.length());
        A.deleteBack();
        System.out.println(A);
        A.clear();
        System.out.println(A.length());
    }*/
}

// Output of this program:
// 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
// 20 19 18 17 16 15 14 13 12 11 10 9 8 7 6 5 4 3 2 1
// 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
// 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
// false
// false
// true
// 1 2 3 4 5 -1 6 7 8 9 11 12 13 14 15 -2 16 17 18 19 20
// 21
// 0

